apiserver/facades/client/application/application.go

```diff
@@ -2199,3 +2275,77 @@ func (api *APIBase) ApplicationsInfo(in params.Entities) (params.ApplicationInfo
 	}
 	return params.ApplicationInfoResults{out}, nil
 }
+
+// lxdCharmProfiler massages a *state.Charm into a LXDProfiler
+// inside of the core package.
+type lxdCharmProfiler struct {
+	Charm Charm
+}
+
+// LXDProfile implements core.lxdprofile.LXDProfiler
+func (p lxdCharmProfiler) LXDProfile() lxdprofile.LXDProfile {
+	if p.Charm == nil {
+		return nil
+	}
+	if profiler, ok := p.Charm.(charm.LXDProfiler); ok {
+		return profiler.LXDProfile()
+	}
+	return nil
+}
+
+// AgentTools is a point of use agent tools requester.
+type AgentTools interface {
+	AgentTools() (*tools.Tools, error)
+}
+
+// ModelConfig is a point of use model configuration.
+type ModelConfig interface {
+	AgentVersion() (version.Number, error)
+}
+
+var (
+	// ErrInvalidAgentVersions is a sentinal error for when we can no longer
+	// upgrade juju using 2.5.x agents with 2.6 or greater controllers.
+	ErrInvalidAgentVersions = errors.Errorf(
+		"Unable to upgrade LXDProfile charms with the current model version. " +
+			"Please run juju upgrade-juju to upgrade the current model to match your controller.")
+)
+
+func getAgentToolsVersion(agentTools AgentTools) (version.Number, error) {
+	tools, err := agentTools.AgentTools()
+	if err != nil {
+		return version.Zero, err
+	}
+	return tools.Version.Number, nil
+}
+
+func getModelConfigVersion(modelConfig ModelConfig) (version.Number, error) {
+	agent, err := modelConfig.AgentVersion()
```

Comment:
> Maybe its just the types/names that need updating? Instead of passing ModelConfig you would pass a modelVersions AgentVersion parameter?

I'll check.

Reply:
Yes you're right, the naming of this is wrong, I'll update and amend.